### ISSUE TITLE
auto_increment needs a persistent sequence

### ISSUE THREAD

**Issue by kostja** [tarantool/tarantool#389]
### Problem

The current implementation of auto_increment **performs an index lookup in the primary key** to find the maximal current value and increment it by one before insert.

This suffers from the following problems: 

* It is impossible to truncate a table but preserve the auto-increment value. Other databases support an independent reset of auto-increment.
* Deleting rows from a table may result in re-using the same values for new rows when they are added
* Lookups are very slow in vinyl.
* concurrent lookups in vinyl may lead to deadlocks and transaction rollbacks. These two reasons render tables with auto-increment nearly useless in vinyl.
* auto-increment is not present in the binary protocol, and using a Lua function for it is a) cumbersome b) requires 'execute' privilege, while auto_increment is a basic data change operation and should not require 'execute'.

### What needs to be done

Auto-increment needs to get a persistent sequence implementation under the hood.

See PostgreSQL for inspiration: 

https://www.postgresql.org/docs/9.5/static/sql-createsequence.html

```
CREATE [ TEMPORARY | TEMP ] SEQUENCE [ IF NOT EXISTS ] name [ INCREMENT [ BY ] increment ]
    [ MINVALUE minvalue | NO MINVALUE ] [ MAXVALUE maxvalue | NO MAXVALUE ]
    [ START [ WITH ] start ] [ CACHE cache ] [ [ NO ] CYCLE ]
    [ OWNED BY { table_name.column_name | NONE } ]
```

We need three system spaces:
- _sequence, for sequence metadata. It lists such properties of the sequence, as start offset, increment, min/max value, the associated space, the definer of the sequence.
- _sequence_data, for actual translation sequence id -> sequence value.
* _space_sequence, for mapping of "automatic" sequences, e.g. sequences used as identity generators for primary keys.

PostgreSQL seems to have over-engineered this quite a bit, but separating sequence object from index object seems to be generally a good idea and fits well into our multiple storage engine architecture - multiple engines don't have to re-implement it again, the best possible implementation is going to look very much like memtx anyway. 

On the complications side, this implies that a sequence, in turn, may get its own definer and access checks - for standalone sequences, at least. We should add owner id to the _sequence definition, but skip access checks for implicitly generated sequences - by definition, if a user has access to a space, they have access to the implicitly generated sequence for this space.

Please also note that PostgreSQL automatically generates and assigns a name to an implicitly generated sequence. This is also a good balance. On one hand, an automatically generated name may conflict with a pre-existing name. On the other, this name is a useful handle to work with the sequence in the future - e.g. reset it independently from the table it stands for, or alter its options.

Useful read:
http://www.neilconway.org/docs/sequences/

### The scope of 1.7.6

In scope of 1.7.6 we need to support only automatically generated sequences. 
**Side note: in the end we support standalone sequences as well. But explicit invocation of nextval() does not work in a cross-engine transaction. nextval() is based on a memtx space, so it works ok in a memtx transaction. automatic sequences work ok for both engines**

The schema for _sequence space is:
```sequence id, owner_id, sequence_name, space id (or nil), increment, stat offset, max value, cycle (yes/no), options```

**Note to the documentaion: check the actual space schema for details**

The schema for _sequence_data is: 
```_sequence_id, sequence_value```
(Index type: hash, for performance reasons).

To enable an implicit sequence one sets a property 'sequence' to true/false/name when creating a primary key. It's only possible to set this property on a key over a single unsigned field.

For example:

```space:create_index('pk', {parts = {1, 'unsigned'}, sequence=true})```

In order to use a sequence from the binary protocol command, pass MP_NIL as the value for the sequence key. If a value bigger than the current sequence value is passed, it automatically promotes the sequence. In the WAL the statement is re-written to contain the actual value.

### Implementation

* A new object type is added to access control - sequence. For implicit sequences, granting a privilege should be banned and is error-prone: e.g. when someone grants someone else a write privilege on a space, they should automatically get a grant on the associated sequence. This is a lot of bookkeeping to do. For implicit sequences, we will skip access checks. For explicit sequences, the user is responsible for the bookkeeping.
* It should be impossible to drop a sequence without first dropping the index it refers to
* dropping the sequence automatically drops all its privileges
* INSERT into a table should update the sequence if a larger value is inserted than is present in the sequence. We may cache the value in ```struct space``` to avoid a lookup in _sequence_data space.
* UPDATE/REPLACE/UPSERT are not an issue since they can not change the primary key
* Vinyl runs the UPDATE in an autonomous transaction
* Memtx may do the same for simplicity reasons - or may use the same transaction for performance reasons; but the default implementation should be simple, not fast.
* Truncate does not affect the sequence by default (PostgreSQL way). This seems to be a more natural default than MySQL one, which resets the sequence. Truncate gets an option to restart the sequence. This is done in a separate transaction for now.
* drop index recursively drops the sequence and all the grants

### What will break

We will keep the old auto_increment command around for compatibility reasons for a while. It will work the old way. To use new automatic sequences, pass mp_nil for the value of the sequence field.

**Comment by rtsisyk** [comment]
@kostja this ticket a breaking change and has high impact on the user experience. It would be nice to implement it in 1.7.3, but we don't have enough resource do to that :(

**Comment by rtsisyk** [comment]
This feature is blocked by #1958, which has been moved to 1.8.

**Comment by kostja** [comment]
Re-scheduling to 1.7.5 in scope of auto-increment support in vinyl.

**Comment by kostja** [comment]
gh-1843 is a duplicate of this bug.

**Comment by rtsisyk** [comment]
I propose minor changes to this spec:

1. Add an unique index to _sequence space to limit the maximum number of sequences attached to a space to 1.

2. Change executeReplace()/executeInsert() to check the indexed field of primary key for NULL
   and replace it by nextval() when field is NULL or update the linked sequence otherwise.
   Multipart and non-numeric primary keys with sequences will not be supported.
   This change can be implemented via virtual `replace_all_keys` implementation to avoid extra costs for existing users.

3. Don't add `auto_increment=true` index option from this spec

4. Don't add a new IPROTO command

5. Deprecate space:auto_increment() method and remove it somewhere in the future.
